src/pyroboplan/core/utils.py

```diff
@@ -356,3 +372,78 @@ def set_collisions(model, collision_model, body1, body2, enable):
                 collision_model.addCollisionPair(pair)
             else:
                 collision_model.removeCollisionPair(pair)
+
+
+def calculate_collision_vector_and_jacobians(
+    model, collision_model, data, collision_data, pair_idx, q
+):
+    """
+    Given collision and distance results from collision model data, computes the collision vector
+    and collision Jacobians at both collision points.
+
+    This is useful for algorithms that perform collision avoidance, such as IK and trajectory optimization.
+
+    Note that forward kinematics, collision, and distance checks must be evaluated first to populate the
+    `data` and `collision_data` variables.
+
+    Parameters
+    ----------
+        model : `pinocchio.Model`
+            The model to use for Jacobian computation.
+        collision_model : `pinocchio.Model`
+            The model to use for collision checking.
+        data : `pinocchio.Data`
+            The model data to use for Jacobian computation.
+        collision_data : `pinocchio.GeometryData`
+            The collision_model data to use for collision checking.
+        pair_idx : int
+            The index of the collision pair from which to extract information.
+        q : array-like
+            The joint configuration of the model.
+
+    Return
+    ------
+        tuple(array-like, array-like, array-like)
+            A tuple containing collision distance vector from frame1 to frame2,
+            and the collision Jacobians at frame1 and frame2.
+    """
+    cp = collision_model.collisionPairs[pair_idx]
+    cr = collision_data.collisionResults[pair_idx]
+    dr = collision_data.distanceResults[pair_idx]
+
+    if cr.isCollision():
+        # According to the HPP-FCL documentation, the normal always points from object1 to object2.
+        contact = cr.getContact(0)
+        coll_points = [
+            contact.pos,
+            contact.pos - contact.normal * contact.penetration_depth,
+        ]
+    else:
+        coll_points = [dr.getNearestPoint1(), dr.getNearestPoint2()]
+
+    distance_vec = coll_points[1] - coll_points[0]
+
+    # Calculate the Jacobians at the parent frames of both collision points.
```

Comment:
super super nit: could be a helper function to compute the jacobian at a point to reduce code duplication

Reply:
I considered this, but this is one of those situations where it makes things even more confusing than it already is, so I chose to leave it there.